### PR TITLE
Fix bug when formatting epoch dates

### LINKED ISSUE #68963: incorrect pagination with date_histogram and format in composite aggregation
ES Version

```
{
  "name" : "baize-server-d7d300ed",
  "cluster_name" : "",
  "cluster_uuid" : "J8pd0v6-SLy_sYY75rVpAQ",
  "version" : {
    "number" : "7.11.0",
    "build_flavor" : "default",
    "build_type" : "",
    "build_hash" : "8ced7813d6f16d2ef30792e2fcde3e755795ee04",
    "build_date" : "2021-02-08T22:44:01.320463Z",
    "build_snapshot" : false,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

Doing a composite aggregation on date_histogram source with `format:epoch_second` param seems to exhibit some inconsistent behavior:

- it can return excess values and "get stuck" on the search_after key.

My testing mapping is:
```
{
  "mappings": {
    "properties": {
      "@timestamp": {
        "format": "yyyy-MM-dd hh:mm:ss",
        "type": "date"
      },
      "app": {
        "type": "keyword"
      },
      "count": {
        "type": "long"
      }
    }
  }
}
```

And data (and use POST _bulk):
```
{ "create" : { "_index" : "test_comp_aggs", "_id" : 1 } }
{ "@timestamp" : "2021-02-14 10:00:00", "app" :  "tiktok",  "count": 1 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 2 } }
{ "@timestamp" : "2021-02-14 10:00:00", "app" :  "wechat",  "count": 1 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 3 } }
{ "@timestamp" : "2021-02-14 10:00:00", "app" :  "facebook",  "count": 1 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 4 } }
{ "@timestamp" : "2021-02-14 10:00:00", "app" :  "wechat",  "count": 1 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 5 } }
{ "@timestamp" : "2021-02-14 10:01:00", "app" :  "wechat",  "count": 2 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 6 } }
{ "@timestamp" : "2021-02-14 10:01:00", "app" :  "facebook",  "count": 2 }
{ "create" : { "_index" : "test_comp_aggs", "_id" : 7 } }
{ "@timestamp" : "2021-02-14 10:01:00", "app" :  "tiktok",  "count": 2 }
```

Running:

1. composite aggregation with a large enough size (10)，and format: epoch_second
```
{
  "size": 0,
  "aggs": {
    "results": {
      "composite": {
        "size": 10,
        "sources": [
            {
              "app": {
                "terms": {
                  "field": "app", 
                  "missing_bucket": true
                }
              }
            }, 
            {
              "ts": {
                "date_histogram": {
                  "field": "@timestamp", 
                  "fixed_interval": "1m", 
                  "time_zone": "Asia/Hong_Kong",
                  "format": "epoch_second"
                }
              }
            }
        ]
      }
    }
  }
}
```

will  return the the correct all (3 in this case) buckets and the latter as the search after key:
```
      "after_key" : {
        "app" : "wechat",
        "ts" : "1613260800"
      },
```

2. then we use after_key from previous result to search again
```{
  "size": 0,
  "aggs": {
    "results": {
      "composite": {
        "after" : {
          "app" : "wechat",
          "ts" : "1613260800"
        },
        "size": 10,
        "sources": [
            {
              "app": {
                "terms": {
                  "field": "app", 
                  "missing_bucket": true
                }
              }
            }, 
            {
              "ts": {
                "date_histogram": {
                  "field": "@timestamp", 
                  "fixed_interval": "1m", 
                  "time_zone": "Asia/Hong_Kong",
                  "format": "epoch_second"
                }
              }
            }
        ]
      }
    }
  }
}
```

will return excess values, and the after_key go stuck:

```
{
  "took" : 0,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 7,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "results" : {
      "after_key" : {
        "app" : "wechat",
        "ts" : "1613260800"
      },
      "buckets" : [
        {
          "key" : {
            "app" : "wechat",
            "ts" : "1613260800"
          },
          "doc_count" : 3
        }
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by not-napoleon**
Resolves #68963

Date based aggregations accept a timezone, which gets applied to both the bucketing logic and the formatter.  This is usually what you want, but in the case of date formats where a timezone doesn't make any sense, it can create problems.  In particular, our formatting logic and our parsing logic were doing different things for `epoch_second` and `epoch_millis` formats with time zones.  This led to a problem on composite where we'd return an after key for the last bucket that would parse to a time before the last bucket, so instead of correctly returning an empty response to indicate the end of the aggregation, we'd keep returning the same last page of data.  

~The proposed fix here is to always coerce the timezone to UTC **for formatters only** when using an epoch-based format.  This aligns with the general expectation that epoch time is always UTC based.~

Turns out we just need to make the formatter timezone aware.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/DocValueFormat.java [hunk 1]
```diff
@@ -229,6 +230,9 @@ public DateTime(DateFormatter formatter, ZoneId timeZone, DateFieldMapper.Resolu
         }
 
         private DateTime(DateFormatter formatter, ZoneId timeZone, DateFieldMapper.Resolution resolution, boolean formatSortValues) {
+            if (formatter.pattern().equals("epoch_second") || formatter.pattern().equals("epoch_millis")) {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/DocValueFormat.java:233]: @pgomulka is there a better way to detect if we have an epoch-based format?  Some kind of `isEpochBased()` method or similar?

File: server/src/main/java/org/elasticsearch/search/DocValueFormat.java [hunk 2]
```diff
@@ -229,9 +229,9 @@ public DateTime(DateFormatter formatter, ZoneId timeZone, DateFieldMapper.Resolu
         }
 
         private DateTime(DateFormatter formatter, ZoneId timeZone, DateFieldMapper.Resolution resolution, boolean formatSortValues) {
-            this.formatter = formatter;
             this.timeZone = Objects.requireNonNull(timeZone);
-            this.parser = formatter.toDateMathParser();
+            this.formatter = formatter.withZone(timeZone);
+            this.parser = this.formatter.toDateMathParser();
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/search/DocValueFormat.java:234]: That's a good one :smile:

**Review by imotov — APPROVED** [review]
LGTM

**Review by pgomulka — APPROVED** [review]
LGTM

**Comment by pgomulka** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/DocValueFormat.java (hunk 2 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/DocValueFormat.java:234]: Credit to Przemyslaw who spotted it.  I was just going to force it to UTC.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**